src/uct/ib/rc/accel/rc_mlx5_ep.c

```diff
@@ -968,80 +966,65 @@ UCS_CLASS_INIT_FUNC(uct_rc_mlx5_ep_t, const uct_ep_params_t *params)
+
+#if IBV_HW_TM
+    if (UCT_RC_MLX5_TM_ENABLED(iface)) {
+        uct_ib_mlx5_iface_put_res_domain(&ep_cleanup_ctx->tm_qp);
+        uct_ib_mlx5_destroy_qp(md, &ep_cleanup_ctx->tm_qp);
     }
-#else
-    (void)uct_ib_mlx5_modify_qp_state(md, qp, IBV_QPS_ERR);
 #endif
 
-    iface->super.rx.srq.available += uct_rc_mlx5_iface_commom_clean(
-            &iface->cq[UCT_IB_DIR_RX],
-            &iface->rx.srq, qp->qp_num);
-
-    /* Synchronize CQ index with the driver, since it would remove pending
-     * completions for this QP (both send and receive) during ibv_destroy_qp().
-     */
-    uct_rc_mlx5_iface_common_update_cqs_ci(iface, &iface->super.super);
-    (void)uct_ib_mlx5_modify_qp_state(md, qp, IBV_QPS_RESET);
-    uct_rc_mlx5_iface_common_sync_cqs_ci(iface, &iface->super.super);
+    uct_ib_mlx5_qp_mmio_cleanup(&ep_cleanup_ctx->qp, ep_cleanup_ctx->reg);
+    uct_ib_mlx5_destroy_qp(md, &ep_cleanup_ctx->qp);
+    ucs_free(ep_cleanup_ctx);
 }
 
-static UCS_CLASS_CLEANUP_FUNC(uct_rc_mlx5_ep_t)
+UCS_CLASS_CLEANUP_FUNC(uct_rc_mlx5_ep_t)
 {
     uct_rc_mlx5_iface_common_t *iface = ucs_derived_of(self->super.super.super.iface,
                                                        uct_rc_mlx5_iface_common_t);
     uct_ib_mlx5_md_t *md              = ucs_derived_of(iface->super.super.super.md,
                                                        uct_ib_mlx5_md_t);
+    uct_rc_mlx5_ep_cleanup_ctx_t *ep_cleanup_ctx;
+
+    ep_cleanup_ctx = ucs_malloc(sizeof(*ep_cleanup_ctx), "ep_cleanup_ctx");
+    ucs_assert_always(ep_cleanup_ctx != NULL);
+    ep_cleanup_ctx->tm_qp = self->tm_qp;
+    ep_cleanup_ctx->qp    = self->tx.wq.super;
+    ep_cleanup_ctx->reg   = self->tx.wq.reg;
 
-    uct_ib_mlx5_txwq_cleanup(&self->tx.wq);
-    uct_rc_mlx5_ep_clean_qp(self, &self->tx.wq.super);
 #if IBV_HW_TM
     if (UCT_RC_MLX5_TM_ENABLED(iface)) {
-        uct_rc_mlx5_ep_clean_qp(self, &self->tm_qp);
-        uct_ib_mlx5_iface_put_res_domain(&self->tm_qp);
         uct_rc_iface_remove_qp(&iface->super, self->tm_qp.qp_num);
-        uct_ib_mlx5_destroy_qp(md, &self->tm_qp);
     }
 #endif
 
     ucs_assert(self->mp.free == 1);
-
-    /* Return all credits if user do flush(UCT_FLUSH_FLAG_CANCEL) before
-     * ep_destroy.
-     */
-    uct_rc_txqp_available_add(&self->super.txqp,
-                              self->tx.wq.bb_max -
-                              uct_rc_txqp_available(&self->super.txqp));
-
-    uct_ib_mlx5_verbs_srq_cleanup(&iface->rx.srq, iface->rx.srq.verbs.srq);
-
-    uct_ib_device_async_event_unregister(&md->super.dev,
-                                         IBV_EVENT_QP_LAST_WQE_REACHED,
-                                         self->tx.wq.super.qp_num);
-    uct_rc_iface_remove_qp(&iface->super, self->tx.wq.super.qp_num);
-    uct_ib_mlx5_destroy_qp(md, &self->tx.wq.super);
+    (void)uct_ib_mlx5_modify_qp_state(md, &self->tx.wq.super, IBV_QPS_ERR);
+    uct_rc_ep_cleanup_qp(&iface->super, &self->super, &ep_cleanup_ctx->super,
```

Comment:
uct_rc_ep_cleanup_qp - parent constructor for cleanup context
add equivalent parent destructor